File: windows/TokenTracker.Windows.Core/Sparkline.cs
using System.Text;

namespace TokenTracker.Windows.Core;

/// Reduces stored history into a compact remaining-% series for one
/// provider/window, downsampled to a menu-friendly width. Pure — no network.
/// Mirrors the Swift <c>SparklineSeries</c>.
public static class SparklineSeries
{
    public static IReadOnlyList<int> Build(
        IReadOnlyList<UsageHistoryEntry> entries,
        Provider provider,
        ForecastWindow window,
        int maxPoints = 20)
    {
        var values = entries
            .OrderBy(entry => entry.RecordedAt)
            .Select(entry => Remaining(entry.Snapshot.Usage(provider), window))
            .Where(value => value is not null)
            .Select(value => value!.Value)
            .ToList();

        if (maxPoints <= 0 || values.Count <= maxPoints)
        {
            return values;
        }

        var result = new List<int>(maxPoints);
        for (var bucket = 0; bucket < maxPoints; bucket++)
        {
            var start = bucket * values.Count / maxPoints;
            var end = Math.Max(start + 1, (bucket + 1) * values.Count / maxPoints);
            result.Add((int)values.GetRange(start, end - start).Average());
        }

        return result;
    }

    private static int? Remaining(ProviderUsage usage, ForecastWindow window) =>
        window == ForecastWindow.FiveHour ? usage.RemainingPercent5h : usage.RemainingPercent7d;
}

/// Renders a 0–100 series as a Unicode block sparkline (absolute scale, so the
/// level and slope are both visible). Shared across platforms — no drawing code.
public static class SparklineText
{
    private static readonly string[] Blocks = { "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█" };

    public static string Render(IReadOnlyList<int> series)
    {
        if (series.Count < 2)
        {
            return string.Empty;
        }

        var builder = new StringBuilder(series.Count);
        foreach (var value in series)
        {
            var clamped = Math.Clamp(value, 0, 100);
            var index = Math.Min(Blocks.Length - 1, clamped * Blocks.Length / 100);
            builder.Append(Blocks[index]);
        }

        return builder.ToString();
    }
}

File: Sources/TokenTrackerCore/Sparkline.swift
import Foundation

/// Reduces stored history into a compact remaining-% series for one
/// provider/window, downsampled to a menu-friendly width. Pure — no network.
public enum SparklineSeries {
    public static func build(
        entries: [UsageHistoryEntry],
        provider: Provider,
        window: ForecastWindow,
        maxPoints: Int = 20
    ) -> [Int] {
        let values = entries
            .sorted { $0.recordedAt < $1.recordedAt }
            .compactMap { remaining($0.snapshot.usage(for: provider), window) }

        guard maxPoints > 0, values.count > maxPoints else { return values }

        var result: [Int] = []
        result.reserveCapacity(maxPoints)
        for bucket in 0..<maxPoints {
            let start = bucket * values.count / maxPoints
            let end = max(start + 1, (bucket + 1) * values.count / maxPoints)
            let slice = values[start..<end]
            result.append(slice.reduce(0, +) / slice.count)
        }
        return result
    }

    private static func remaining(_ usage: ProviderUsage, _ window: ForecastWindow) -> Int? {
        switch window {
        case .fiveHour: return usage.remainingPercent5h
        case .sevenDay: return usage.remainingPercent7d
        }
    }
}

/// Renders a 0–100 series as a Unicode block sparkline (absolute scale, so the
/// level and slope are both visible). Shared across platforms — no drawing code.
public enum SparklineText {
    private static let blocks = ["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"]

    public static func render(_ series: [Int]) -> String {
        guard series.count >= 2 else { return "" }
        return series.map { value in
            let clamped = min(100, max(0, value))
            let index = min(blocks.count - 1, clamped * blocks.count / 100)
            return blocks[index]
        }.joined()
    }
}
